File: src/main/resources/META-INF/resources/frontend/google-map.js
/** 
@license
Copyright 2010-2019 Neotropic SAS <contact@neotropic.co>.

Licensed under the Apache License, Version 2.0 (the "License"); 
you may not use this file except in compliance with the License. 
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software 
distributed under the License is distributed on an "AS IS" BASIS, 
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
See the License for the specific language governing permissions and 
limitations under the License.
*/

import {html, PolymerElement} from '@polymer/polymer/polymer-element.js';
import { FlattenedNodesObserver } from '@polymer/polymer/lib/utils/flattened-nodes-observer.js';
import {MapApi} from './google-map-api.js';
import * as Constants from './google-map-constants.js';
import {GoogleMapMarker} from './google-map-marker.js';
import {GoogleMapPolyline} from './google-map-polyline.js';
/**
 * `google-map`
 * &lt;google-map&gt; is a web component displays a map using Maps JavaScript API
 *
 * @customElement
 * @polymer
 * @demo demo/index.html
 * @extends PolymerElement
 * @summary Custom Element for Google Map
 * @author Johny Andres Ortega Ruiz {@literal <johny.ortega@kuwaiba.org>}
 */
class GoogleMap extends PolymerElement {
  
  static get is() {
    return Constants.googleMap;
  }
    
  static get template() {
    return html`
      <style>
        /* Always set the map height explicitly to define the size of the div
        * element that contains the map. */
        #map {          
          height: 100%;
          width: 100%;
        }       
      </style>
      <!--Container for the map-->
      <div id="map"></div>
      <slot></slot>
      <!--<slot id="markers"></slot>-->
    `;
  }

  static get properties() {
    return {
      /**
       * The string contains your application API key. See https://developers.google.com/maps/documentation/javascript/get-api-key
       */
      apiKey: {
        type: String,
        value: ''
      },
      /**
       * Specifies a client Id
       */
      clientId: String,
      /**
       * The name of the additional library or libraries to load
       * drawing,geometry,places,visualization
       */
      libraries: String,
      /**
       * The initial map center latitude coordinate ranges between [-90, 90] degrees 
       */
      lat: {
        type: Number,
        value: 2.4573831,
        observer: '_latChanged'
      },
      /**
       * The initial map center longitude coordinate ranges between [-180, 180] degress
       */
      lng: {
        type: Number,
        value: -76.6699746,
        observer: '_lngChanged'
      },
      /**
       * The initial map zoom level, specify the value as an integer
       */
      zoom: {
        type: Number,
        value: 10,
        observer: '_zoomChanged'
      },
      /**
       * hybrid, roadmap, satellite, terrain
       */
      mapTypeId: {
        type: String,
        value: 'roadmap',
        observer: '_mapTypeIdChanged'
      }
    };
  }
  /*
  constructor() {
    super();
  }
  */
  /*
  connectedCallback() {
    super.connectedCallback();
  }
  */
  disconnectedCallback() {
    super.disconnectedCallback();
  }
  
  /**
   * Display a map
   * @return {void}
   */
  initMap() {    
    /**
     * Map instance which define a single map on a page
     * @type google.map.Map
     */
    this.map = new google.maps.Map(this.shadowRoot.querySelector('#map'), {
      center: {lat: this.lat, lng: this.lng},
      zoom: this.zoom
    });
    this.map.setMapTypeId(this.mapTypeId);
    /*
    Events:
    bounds_changed, 
    *center_changed, 
    *click, 
    *dblclick, 
    drag, 
    dragend, 
    dragstart, 
    heading_changed, 
    idle, 
    maptypeid_changed, 
    *mousemove, 
    *mouseout, 
    *mouseover, 
    projection_changed, 
    *rightclick, 
    tilesloaded, 
    tilt_changed, 
    *zoom_changed
    */
    var _this = this;
    this.map.addListener('center_changed', function() {
      _this.lat = _this.map.getCenter().lat();
      _this.lng = _this.map.getCenter().lng();
      _this.dispatchEvent(new CustomEvent('map-center-changed'));
    });
    this.map.addListener('mousemove', function(event) {
      _this.dispatchEvent(new CustomEvent('map-mouse-move'));
    });
    this.map.addListener('mouseout', function(event) {
      _this.dispatchEvent(new CustomEvent('map-mouse-out'));
    });
    this.map.addListener('mouseover', function(event) {
      _this.dispatchEvent(new CustomEvent('map-mouse-over'));
    });
    this.map.addListener('zoom_changed', function() {
      _this.zoom = _this.map.getZoom();
      _this.dispatchEvent(new CustomEvent('map-zoom-changed'));
    });
    this.map.addListener('click', function(event) {
      var lat = event.latLng.lat();
      var lng = event.latLng.lng();
      _this.dispatchEvent(new CustomEvent('map-click', 
        {detail: {lat: lat, lng: lng}}));
    });
    this.map.addListener('dblclick', function(event) {
      _this.dispatchEvent(new CustomEvent('map-dbl-click'));
    });
    this.map.addListener('rightclick', function(event) {
      var lat = event.latLng.lat();
      var lng = event.latLng.lng();
      _this.dispatchEvent(new CustomEvent('map-right-click', 
        {detail: {lat: lat, lng: lng}}));
    });
    var _this = this;
    this._observer = new FlattenedNodesObserver(this, (info) => {      
      _this._processAddedNodes(info.addedNodes);
      _this._processRemovedNodes(info.removedNodes);
    });
  }
  /**
   * Loads the Maps JavaScript API
   * when the api is ready, it will call this.initMap()
   * @override
   * @return {void}
   */
  ready() {
    super.ready();    

    const mapApi = new MapApi(this.apiKey, this.clientId, this.libraries);
    mapApi.load().then(() => {this.initMap()});
  }
       
  _processAddedNodes(addedNodes) {
    for (var i = 0; i < addedNodes.length; i++) {
      if (addedNodes[i].localName === Constants.googleMapMarker) {
        addedNodes[i].draw(this.map);
      }
      if (addedNodes[i].localName === Constants.googleMapPolyline) {
        addedNodes[i].draw(this.map);
      }
    }
  }

  _processRemovedNodes(removedNodes) {
    for (var i = 0; i < removedNodes.length; i++) {
      if (removedNodes[i].localName === Constants.googleMapMarker) {
        removedNodes[i].remove();
      }
      else if (removedNodes[i].localName === Constants.googleMapPolyline) {
        removedNodes[i].remove();
      }
    }
  }

  _latChanged(newValue, oldValue) {
    if (this.map !== undefined && this.map.getCenter() !== undefined && 
      this.map.getCenter().lat() !== newValue) {
      this.map.setCenter({lat: newValue, lng: this.map.getCenter().lng()});
    }
  }

  _lngChanged(newValue, oldValue) {
    if (this.map !== undefined && this.map.getCenter() !== undefined && 
      this.map.getCenter().lng() !== newValue) {
      this.map.setCenter({lat: this.map.getCenter().lat(), lng: newValue});
    }
  }

  _zoomChanged(newValue, oldValue) {
    if (this.map !== undefined && 
      this.map.getZoom() !== newValue) {
      this.map.setZoom(newValue);
    }
  }

  _mapTypeIdChanged(newValue, oldValue) {
    if (this.map !== undefined && 
      this.map.getMapTypeId() !== newValue) {
      this.map.setMapTypeId(newValue);
    }
  }
}

window.customElements.define(GoogleMap.is, GoogleMap);

File: src/main/java/com/neotropic/vaadin14/component/googlemap/GoogleMap.java
/*
 *  Copyright 2010-2019 Neotropic SAS <contact@neotropic.co>.
 *  
 *  Licensed under the Apache License, Version 2.0 (the "License"); 
 *  you may not use this file except in compliance with the License. 
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software 
 *  distributed under the License is distributed on an "AS IS" BASIS, 
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
 *  See the License for the specific language governing permissions and 
 *  limitations under the License.
 */
package com.neotropic.vaadin14.component.googlemap;

import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.ComponentEventListener;
import com.vaadin.flow.component.Synchronize;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.shared.Registration;

/**
 *
 * @author Johny Andres Ortega Ruiz {@literal <johny.ortega@kuwaiba.org>}
 */
@Tag("google-map")
@JsModule("./google-map.js")
public class GoogleMap extends Component {
    /**
     * @param apiKey Your API key and set the client id to null. See https://developers.google.com/maps/documentation/javascript/get-api-key
     * @param clientId Your client id and set the apiKey to null. See https://developers.google.com/maps/documentation/javascript/get-api-key
     */    
    public GoogleMap(String apiKey, String clientId) {
        getElement().getStyle().set(Constants.Property.WIDTH, "100%");
        getElement().getStyle().set(Constants.Property.HEIGHT, "100%");        
        getElement().setProperty(Constants.Property.API_KEY, apiKey);
        getElement().getStyle().set(Constants.Property.MIN_WIDTH, "250px");  
        getElement().setProperty(Constants.Property.CLIENT_ID, clientId);
    }
    public GoogleMap(String apiKey, String clientId, String width, String height) {
        getElement().getStyle().set(Constants.Property.WIDTH, width);
        getElement().getStyle().set(Constants.Property.HEIGHT, height);  
        getElement().getStyle().set(Constants.Property.MIN_WIDTH, "250px");  
        getElement().setProperty(Constants.Property.API_KEY, apiKey);
        getElement().setProperty(Constants.Property.CLIENT_ID, clientId);
    }
    @Synchronize(property = "lat", value = "map-center-changed")
    public double getCenterLat() {
        return getElement().getProperty(Constants.Property.LAT, Constants.Default.LAT);
    }
    
    public void setCenterLat(double lat) {
        getElement().setProperty(Constants.Property.LAT, lat);
    }
    @Synchronize(property = "lng", value = "map-center-changed")
    public double getCenterLng() {
        return getElement().getProperty(Constants.Property.LNG, Constants.Default.LNG);
    }
    
    public void setCenterLng(double lng) {
        getElement().setProperty(Constants.Property.LNG, lng);
    }
    @Synchronize(property = "zoom", value = "map-zoom-changed")
    public double getZoom() {
        return getElement().getProperty(Constants.Property.ZOOM, Constants.Default.ZOOM);
    }
    
    public void setZoom(double zoom) {
        getElement().setProperty(Constants.Property.ZOOM, zoom);
    }
    
    public String getMapTypeId() {
        return getElement().getProperty(Constants.Property.MAP_TYPE_ID, Constants.Default.MAP_TYPE_ID);
    }
    
    public void setMapTypeId(String mapTypeId) {
        getElement().setProperty(Constants.Property.MAP_TYPE_ID, mapTypeId);
    }
    
    
    public void setWidth(String width) {
        getElement().setProperty(Constants.Property.WIDTH, width);
    }
 
    public void setHeight(String height) {
        getElement().setProperty(Constants.Property.HEIGHT, height);
    }
    
    public void setMinWidth(String width) {
        getElement().setProperty(Constants.Property.MIN_WIDTH, width);
    }
 
    public void setMinHeight(String height) {
        getElement().setProperty(Constants.Property.MIN_HEIGHT, height);
    }
    
    public void newMarker(GoogleMapMarker googleMapMarker) {
        getElement().appendChild(googleMapMarker.getElement());
    }
    
    public void removeMarker(GoogleMapMarker googleMapMarker) {
        getElement().removeChild(googleMapMarker.getElement());
    }
    
    public void newPolyline(GoogleMapPolyline polyline) {
        getElement().appendChild(polyline.getElement());
    }
    
    public void removePolyline(GoogleMapPolyline polyline) {
        getElement().removeChild(polyline.getElement());
    }
            
    public Registration addMapClickListener(ComponentEventListener<GoogleMapEvent.MapClickEvent> listener) {
        return addListener(GoogleMapEvent.MapClickEvent.class, listener);
    }
    
    public Registration addMapDblClickListener(ComponentEventListener<GoogleMapEvent.MapDblClickEvent> listener) {
        return addListener(GoogleMapEvent.MapDblClickEvent.class, listener);
    }
    
    public Registration addMapRightClickListener(ComponentEventListener<GoogleMapEvent.MapRightClickEvent> listener) {
        return addListener(GoogleMapEvent.MapRightClickEvent.class, listener);
    }
    
    public Registration addMapCenterChangedListener(ComponentEventListener<GoogleMapEvent.MapCenterChangedEvent> listener) {
        return addListener(GoogleMapEvent.MapCenterChangedEvent.class, listener);
    }
    
    public Registration addMapMouseMoveListener(ComponentEventListener<GoogleMapEvent.MapMouseMoveEvent> listener) {
        return addListener(GoogleMapEvent.MapMouseMoveEvent.class, listener);
    }
    
    public Registration addMapMouseOutListener(ComponentEventListener<GoogleMapEvent.MapMouseOutEvent> listener) {
        return addListener(GoogleMapEvent.MapMouseOutEvent.class, listener);
    }
    
    public Registration addMapMouseOverListener(ComponentEventListener<GoogleMapEvent.MapMouseOverEvent> listener) {
        return addListener(GoogleMapEvent.MapMouseOverEvent.class, listener);
    }
    
    public Registration addMapZoomChangedListener(ComponentEventListener<GoogleMapEvent.MapZoomChangedEvent> listener) {
        return addListener(GoogleMapEvent.MapZoomChangedEvent.class, listener);
    }
}

